volatility3/framework/__init__.py

```
@@ -52,6 +53,32 @@ def require_interface_version(*args) -> None:
                 )
 
 
+class Deprecation:
+    """Deprecation related methods."""
+
+    @staticmethod
+    def deprecated_method(replacement: Callable, additional_information: str = ""):
+        """A decorator for marking functions as deprecated.
+
+        Args:
+            replacement: The replacement function overriding the deprecated API, in the form of a Callable (typically a method)
+            additional_information: Information appended at the end of the deprecation message
+        """
+
+        def decorator(deprecated_func):
+            @functools.wraps(deprecated_func)
+            def wrapper(*args, **kwargs):
+                nonlocal replacement, additional_information
+                deprecation_msg = f"Method \"{deprecated_func.__module__ + '.' + deprecated_func.__name__}\" is deprecated, use \"{replacement.__module__ + '.' + replacement.__name__}\" instead. {additional_information}"
```

Can we include a test here of the callable's parent class, verify it's versionable and then check the version (if a version's been provided to the decorator)?  If that's two hard then passing in the class and the method name could work.  That way, it can throw an exception rather than just a warning if the version is suitably wrong...

Also, this shouldn't use vollog, it should use `warnings.warn(f"{name} is deprecated", FutureWarning)` or similar, since Deprecation warnings only show up once whereas normal log warnings may show up 100 times if called 100 times in the code (in a loop for example)...

---

Thanks for pointing out `warnings.warn`, I was also concerned about the message going out a lot.

I managed to grab the base class from the method, make sanity checks and finally verify the version requirement. I didn't add a `DeprecatedException` kind of thing, as `UnsatisfiedException` will be generic and accomplish the nice backtrace needed to debug and patch the issue.

---

Example:

```
An unsatisfied framework exception was encountered post plugin construction:

Unsatisfied requirement volatility3.framework.symbols.linux.lookup_module_address.Modules: Version 1.2.0 dependency on volatility3.framework.symbols.linux.utilities.modules.Modules unmet
Unable to validate the requirements: ['volatility3.framework.symbols.linux.lookup_module_address.Modules']
```